cranelift-codegen: Add support for immediate to memory moves in x64

saulecabrera opened the pull request:
This change adds support for immediate to memory moves in x64 which are needed by Winch for zeroing local slots.

This change follows the guideline in `isa/x64/inst/emit` and uses other instructions (immediate to register moves) as a base for the test cases.

The instruction encoding expectation was derived by assembling each instruction and inspecting the assembly with `objdump`.

<!--

Please ensure that the following steps are all taken care of before submitting
the PR.

- [ ] This has been discussed in issue #..., or if not, please tell us why
  here.
- [ ] A short description of what this does, why it is needed; if the
  description becomes long, the matter should probably be discussed in an issue
  first.
- [ ] This PR contains test cases, if meaningful.
- [ ] A reviewer from the core maintainer team has been assigned for this PR.
  If you don't know who could review this, please indicate so. The list of
  suggested reviewers on the right can help you.

Please ensure all communication adheres to the [code of
conduct](https://github.com/bytecodealliance/wasmtime/blob/master/CODE_OF_CONDUCT.md).
-->

cfallin commented on cranelift/codegen/src/isa/x64/inst/emit_tests.rs:
```diff
@@ -2560,6 +2560,57 @@ fn test_x64_emit() {
+        "48C7456300000080",
+        "movq    $-2147483648, 99(%rbp)",
+    ));
+
```
Can we add `Size16` and `Size8` cases here too?

saulecabrera replied:
Done, thanks!

cfallin commented on cranelift/codegen/src/isa/x64/inst/emit.rs:
```diff
@@ -670,6 +670,34 @@ pub(crate) fn emit(
+            let bytes = size.to_bytes();
+
+            let (opcode, rex, size) = match *size {
+                OperandSize::Size8 => (0xC6, default_rex, bytes),
```
Add a comment here perhaps that we *don't* need `always_emit_if_8bit_needed()` on the REX flags, because the dst is a memory operand, not a possibly 8-bit register?

saulecabrera replied:
Agreed, done!

cfallin reviewed (approved):
LGTM, just a request for a few more tests below and maybe a comment; thanks!

saulecabrera commented:
@cfallin thanks for the review; Im requesting a re-review since aside from addressing your comments, I uncovered a bug in the 16-bit case, in which the operand size override (`LegacyPrefixes::_66`) was missing, so I ended up refactoring the emission to take this into account, aside from that everything else is the same.

cfallin commented on cranelift/codegen/src/isa/x64/inst/emit.rs:
```diff
@@ -670,6 +670,38 @@ pub(crate) fn emit(
+            };
+
+            // 8-bit C6 /0 ib
+            // 16-bit C7 /0 iw
```
Can we update this comment to include the prefix as well?

saulecabrera replied:
Good catch, thanks! Updated!

cfallin reviewed (approved):
LGTM modulo one comment tweak below -- thanks!
